go/teams/chain.go

```diff
@@ -957,6 +1004,87 @@ func (t *TeamSigChainPlayer) inflateLinkHelper(
 	return &iRes.newState, nil
 }
 
+// sanityCheckInvites sanity checks a raw SCTeamInvites section and coerces it into a
+// format that we can use. It checks:
+//  - that invite IDs are repeated
+//  - that <name,type> pairs aren't reused
+//  - that IDs parse into proper keybase1.TeamInviteIDs
+//  - that the invite type parses into proper TeamInviteType, or that it's an unknown
+//    invite that we're OK to not act upon.
+// Returns nicely formatted data structures.
+func (t *TeamSigChainPlayer) sanityCheckInvites(invites SCTeamInvites) (additions map[keybase1.TeamRole][]keybase1.TeamInvite, cancelations []keybase1.TeamInviteID, err error) {
+
+	type assignment struct {
+		i    SCTeamInvite
+		role keybase1.TeamRole
+	}
+	var all []assignment
+
+	if invites.Admins != nil {
+		additions[keybase1.TeamRole_ADMIN] = nil
+		for _, i := range *invites.Admins {
+			all = append(all, assignment{i, keybase1.TeamRole_ADMIN})
+		}
+	}
+
+	if invites.Writers != nil {
+		additions[keybase1.TeamRole_WRITER] = nil
+		for _, i := range *invites.Writers {
+			all = append(all, assignment{i, keybase1.TeamRole_WRITER})
+		}
+	}
+
+	if invites.Readers != nil {
+		additions[keybase1.TeamRole_READER] = nil
+		for _, i := range *invites.Readers {
+			all = append(all, assignment{i, keybase1.TeamRole_READER})
+		}
+	}
+
+	// Set to `true` if it was an addition and `false` if it was a deletion
+	byID := make(map[keybase1.TeamInviteID]bool)
+	byName := make(map[string]bool)
+
+	keyFunc := func(i SCTeamInvite) string {
+		return fmt.Sprintf("%s:%s", i.Type, i.Name)
+	}
+
+	if invites.Cancel != nil {
+		for _, c := range *invites.Cancel {
+			id, err := c.TeamInviteID()
+			if err != nil {
+				return nil, nil, err
+			}
+			if byID[id] {
+				return nil, nil, NewInviteError(fmt.Sprintf("ID %s appears twice as a cancelation", c))
+			}
+			byID[id] = false
+			cancelations = append(cancelations, id)
+		}
+	}
+
+	for _, invite := range all {
+		res, err := invite.i.TeamInvite(invite.role)
+		if err != nil {
+			return nil, nil, err
+		}
+		id := res.Id
+		_, seen := byID[id]
+		if seen {
+			return nil, nil, NewInviteError(fmt.Sprintf("Invite ID %s appears twice in invite set", id))
```

Comment:
This checks within the link, but not whether it collides with previous links. So they will just get overridden. This probably won't cause any problems, but we should decide on the rules and at least have the server enforce them.